coalib/bearlib/aspects/base.py

```diff
@@ -1,8 +1,94 @@
+            parent = obj
+            parent_qualname = type(obj).__qualname__
+        else:
+            parent = owner
+            parent_qualname = owner.__qualname__
+
+        def get_subaspect(subaspect):
+            """
+            Get a subaspect from an aspectclass or aspectclass instance.
+
+            >>> import coalib.bearlib.aspects as aspects
+            >>> metadata = aspects['Metadata']
+            >>> commit_msg = aspects['CommitMessage']
+            >>> shortlog = aspects['Shortlog']
+
+            We can get direct children.
+
+            >>> metadata.get(commit_msg)
+            <aspectclass 'Root.Metadata.CommitMessage'>
+
+            Or even a grandchildren.
+
+            >>> metadata.get(shortlog)
+            <aspectclass 'Root.Metadata.CommitMessage.Shortlog'>
+
+            Trying to access children of aspect instance will raise
+            NotImplementedError because of instanced aspect doesn't instance
+            its children and will cause wrong result.
+            See https://github.com/coala/coala/issues/4388
+
+            >>> metadata('Python').get(commit_msg)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Cannot access children of aspect instance.
+
+            But, passing subaspect instance as argument is prohibited, because
+            it doesn't really make sense.
+
+            >>> metadata('Python').get(commit_msg('Java'))
+            Traceback (most recent call last):
+            ...
+            AttributeError: Cannot search an aspect instance using another
+            aspect instance as argument.
+
+            :param subaspect: An subaspect that we want to find in an
+                              aspectclass.
+            :return:          An aspectclass. Return None if not found.
+            """
+            # Avoid circular import
+            from .meta import issubaspect
+            if isinstance(subaspect, str):
+                subaspect = coalib.bearlib.aspects[subaspect]
+            if not issubaspect(subaspect, parent):
+                return None
+            if isinstance(subaspect, aspectbase):
+                raise AttributeError('Cannot search an aspect instance using '
+                                     'another aspect instance as argument.')
+            if isinstance(parent, aspectbase):
+                raise NotImplementedError('Cannot access children of aspect '
+                                          'instance.')
+            if issubaspect(parent, subaspect):
+                return parent
+            # Trim common parent name
+            aspect_path = re.sub(parent_qualname + '.', '',
+                                 subaspect.__qualname__)
+            aspect_path = aspect_path.split('.')
+            child = parent
+            # Traverse through children until we got our subaspect
+            for path in aspect_path:
+                child = child.subaspects[path]
+            return child
```

Comment:
You can do this whole search much simpler! You don't need a tree traversal at all, since every aspectclass has a `.parent` attribute. Just go from the `subaspect` back until you find the `parent` or not :)